FIPS/fip-0032.md

```diff
@@ -21,290 +21,426 @@ replaces: N/A
+Wasm memory.
 
-- `get_chain_randomness`: `<<TODO>>`
-- `get_beacon_randomness`: `<<TODO>>`
-- `verify_consensus_fault`: `<<TODO>>`. This is a fixed cost that covers all blockstore operations,
-  signature verifications, and other logic needed to resolve whether a consensus fault has occured.
-  Note that differs from network v15 and earlier, where the cost is variable depending on the exact
-  blockstore operations performed.
-- Blockstore operations: already absorbed in the fees specified above ("IPLD
-  state management fees").
+The pricing formula is:
+
+```
+op_ipld_blockread := p_syscall_gas + (p_memcpy_gas_per_byte * byte_length)
+```
+
+**Pricing for `ipld::block_write`**
+
+_Overhead:_ This syscall performs an memcpy from Wasm memory into the FVM
+kernel's buffers, with a corresponding alloc (absorbed into the memcpy cost).
+The kernel buffers are retained until the actor call finishes.
+
+The pricing formula is:
+
+```
+op_ipld_blockwrite := p_syscall_gas + (p_memcpy_gas_per_byte * byte_length)
+```
+
+**Pricing for `ipld::block_link`**
+
+_Overhead:_ This syscall incurs a hashing operation to compute the CID of the
+block. It then copies it over to the buffered blockstore, where it is retained
+until the Machine is flushed and the blocks are written to the Node's blockstore
+via an extern. Because the extern is invoked by the system, the actor is not
+charged for it.
+
+The pricing formula for compute gas is:
+
+```
+p_blocklink_base_gas := 353640    # extant fee
+p_storage_gas_multiplier := 1300  # extant fee
+op_ipld_blocklink_compute := p_syscall_gas + p_blocklink_base_gas + (p_memcpy_gas_per_byte * byte_length * 2)
+op_ipld_blocklink_storage := p_storage_gas_multiplier * byte_length
+op_ipld_blocklink := op_ipld_blocklink_compute + op_ipld_blocklink_storage
+```
+
+**Pricing for `ipld::block_stat`**
+
+No relevant overhead.
+
+The pricing formula is:
+
+```
+op_ipld_blockstat := p_syscall_gas
+```
 
 ### Under/overflow handling
 
 Operations that have the potential of either underflowing or overflowing should be performed
 using [saturation arithmetic](https://en.wikipedia.org/wiki/Saturation_arithmetic). The valid
-range for any fuel operation, including total fuel consumed, is [0, 2<sup>64</sup>-1]. 
+range for any execution units operation, including total execution units consumed, is [0, 2<sup>64</sup>-1]. 
 
-The valid range for any gas operation, including gas available and total gas used, is 
-[0, 2<sup>63</sup>-1]. Gas operations only go up to 2^63-1 because they are performed using
-signed arithmetic. This is because the Filecoin protocol has some gas costs that are negative as 
-of nv15.
+The valid range for any gas operation, including gas available and total gas
+used, is [0, 2<sup>63</sup>-1]. Gas operations only go up to 2^63-1 because they
+are performed using signed arithmetic. This is because the Filecoin protocol has
+some gas costs that are negative as of nv15.
 
```

Comment:
This section isn't entirely accurate:

1. At the upper end, we now saturate to `(2^63-1)/1000`, because we track milligas. But this is just an implementation detail. If we end up saturating here, we'll just end up running out of gas.
2. At the lower end, we _can_ go negative. But we'll floor at zero when we finish executing.